src/main/java/drukier/maze/Maze.java

```diff
@@ -1,170 +1,222 @@
+                if (!visit.isEmpty()) {
+                    current = visit.pop();
+                }
+                else{
+                    System.out.println("visitedCells = " + visitedCells);
+                }
             }
-
         }
-        return array;
-
     }
 
+    //if I want to start randomly
     private MazeCell findStartCell() {
-        Random currentX = new Random();
-        Random currentY = new Random();
+        MazeCell start;
 
-        int cellX = currentX.nextInt(mazeHeight - 1);
-        int cellY = currentY.nextInt(mazeWidth - 1);
+        int cellX = random.nextInt(mazeWidth - 1);
+        int cellY = random.nextInt(mazeHeight - 1);
 
-        start = new MazeCell(cellX, cellY);
+        start = getMazeCell(cellX, cellY);
         return start;
 
     }
 
+    private MazeCell getMazeCell(int cellX, int cellY) {
+        return maze[cellX][cellY];
+    }
 
-    private MazeCell getNextCell(MazeCell cell) {
+    private boolean nextCell(MazeCell current) {
+
+        MazeCell nextCell = null;
+        int nextX = current.getCellX();
+        int nextY = current.getCellY();
+        boolean success = false;
+        direction = getDirection();
+        int doneDirection = 0;
+
+        while (nextCell == null && doneDirection < 4) {
+
+            switch (direction) {
+                case UP:
+                    nextY--;
+                    doneDirection++;
+                    break;
+                case DOWN:
+                    nextY++;
+                    doneDirection++;
+                    break;
+                case RIGHT:
+                    nextX++;
+                    doneDirection++;
+                    break;
+                case LEFT:
+                    nextX--;
+                    doneDirection++;
+                    break;
+            }
+
+            if ((mazeContains(nextX, nextY)) && (!(getMazeCell(nextX, nextY).isVisited()))) {
 
-        ArrayList<MazeCell> neighbors = null;
+                    nextCell = getMazeCell(nextX, nextY);
+                    current.setNext(nextCell);
+                    removeWalls(current, nextCell, direction);
+                    success = true;
+                }
+                //reset current and get next direction
+                else {
+                    nextX = current.getCellX();
+                    nextY = current.getCellY();
+                    direction = nextDirection();
+                }
 
-        MazeCell top = null;
-        MazeCell bottom = null;
-        MazeCell left = null;
-        MazeCell right = null;
 
 
-        if ((cell.getCellX() - 1) > 0 && (cell.getCellX() - 1) < mazeHeight) {
-            top = new MazeCell(cell.getCellX() - 1, cell.getCellY());
         }
+        if (!current.isVisited()) {
+            visitedCells++;
+            current.setVisited(true);
+        }
+        return success;
+    }
+
+    private boolean mazeContains(int x, int y) {
+        boolean valid = false;
```

Comment:
Whoever told you that methods are only allowed one return statement is wrong.
Also this is easier written as 
`return (x < mazeWidth) && (y < mazeHeight) && (x >= 0) && (y >= 0);`

Reply:
thank you

Reply:
I took it apart to debug and forgot to go back